Lilu/Headers/kern_util.hpp

```diff
@@ -944,4 +945,285 @@ class evector_base {
+	 *
+	 *  @return `true` if the buffer is full, `false` otherwise.
+	 */
+	inline bool isFull() {
+		IORecursiveLockLock(lock);
+		bool retVal = (count == size) && (indexr == indexw);
+		IORecursiveLockUnlock(lock);
+		return retVal;
+	}
+	
+	/**
+	 *  Get the number of elements in the circular buffer
+	 *
+	 *  @return The current number of elements in the buffer.
+	 */
+	inline IOItemCount getCount() {
+		IORecursiveLockLock(lock);
+		IOItemCount retVal = count;
+		IORecursiveLockUnlock(lock);
+		return retVal;
+	}
+	
+	/**
+	 *  Write the given element to the circular buffer
+	 *
+	 *  @param element The element to write
+	 *  @return `true` on success, `false` if the buffer is full.
+	 */
+	inline bool write(const T &element) {
+		if (isFull())
+			return false;
+		
+		IORecursiveLockLock(lock);
+		storage[indexw] = element;
+		indexw += 1;
+		indexw %= size;
+		count += 1;
+		IORecursiveLockUnlock(lock);
+		return true;
+	}
+	
+	/**
+	 *  Read the next element from the circular buffer
+	 *
+	 *  @param element The element read from the buffer
+	 *  @return `true` on success, `false` if the buffer is empty.
+	 */
+	inline bool read(T& element) {
+		if (isEmpty())
+			return false;
+		
+		IORecursiveLockLock(lock);
+		element = storage[indexr];
+		indexr += 1;
+		indexr %= size;
+		count -= 1;
+		IORecursiveLockUnlock(lock);
+		return true;
+	}
+};
+
+/**
+ *  Wrap an object that is not an instance of OSObject
+ */
+class OSObjectWrapper: public OSObject {
+	/**
+	 *  Constructors & Destructors
+	 */
+	OSDeclareDefaultStructors(OSObjectWrapper);
+	
+	using super = OSObject;
+	
+	/**
+	 *  Wrapped object
+	 */
+	void *object;
+	
+public:
+	/**
+	 *  Initialize the wrapper with the given object
+	 *
+	 *  @param object The wrapped object that is not an `OSObject`
+	 *  @return `true` on success, `false` otherwise.
+	 */
+	EXPORT bool init(void *object);
+	
+	/**
+	 *  Reinterpret the wrapped object as the given type
+	 *
+	 *  @return The wrapped object of the given type.
+	 */
+	template <typename T>
+	T *reinterpretWrappedObjectAs() {
+		return reinterpret_cast<T*>(object);
+	}
+	
+	/**
+	 *  Create a wrapper for the given object that is not an `OSObject`
+	 *
+	 *  @param object A non-null object
+	 *  @return A non-null wrapper on success, `nullptr` otherwise.
+	 *  @warning The caller is responsbile for managing the lifecycle of the given object.
+	 */
+	EXPORT static OSObjectWrapper *of(void *object);
+};
+
+namespace Value {
+	template <typename T>
+	struct Value {
+		T value;
+
+		explicit Value(T value) : value(value) {}
+
+	#if __cplusplus >= 201703L
+		// Available as of C++17
+		template<typename... Ts>
+		bool isOneOf(Ts... args) {
+			return ((value == args) || ...);
+		}
+
+		// Available as of C++17
+		template <typename... Ts>
+		bool isNotOneOf(Ts... args) {
+			return ((value != args) && ...);
+		}
+	#endif
+	};
+	
+	template <typename T>
+	static Value<T> of(T value) {
+		return Value<T>(value);
```

Comment:
Do we really want copying here? To be honest, I do not quite understand why do we need to wrap objects in this class to do isOneOf comparison. Cannot we just make a function that will take an object and a list of values to compare? Without any kind of construction/object wrapping. Feels much cleaner.

Reply:
Well, based on my observation, the compiler usually optimize away the copy construction. I changed it to a constant reference to the value.

> Cannot we just make a function that will take an object and a list of values to compare?

Yes, we could like the one below.

```
template <typename T, typename... Ts>
bool isOneOf(T value, Ts... args) { ... }

// Invocation:
isOneOf(1, 2, 5, 1, 8);
Value::of(1).isOneOf(2, 5, 1, 8);
```

But since the function is variadic, I personally find it better to separate the object of interest from the list of targets to improve the code readability. i.e. `if (Value::of(1).isOneOf(2, 5, 1, 8))` reads like a sentence. What do you think?

Reply:
I guess there is something reasonable in your words, I am ok.